src/include/pka_cpu.h

```diff
@@ -80,14 +86,191 @@ static inline uint64_t pka_early_cpu_speed(void)
 }
 #else
 
+#ifdef __BIG_ENDIAN__
+#define WORD(x) (u16)((x)[0] + ((x)[1] << 8))
+#define DWORD(x) (u32)((x)[0] + ((x)[1] << 8) + ((x)[2] << 16) + ((x)[3] << 24))
+#define QWORD(x) (U64(DWORD(x), DWORD(x + 4)))
+#else
+#define WORD(x) (uint16_t)(*(const uint16_t *)(x))
+#define DWORD(x) (uint32_t)(*(const uint32_t *)(x))
+#define QWORD(x) (*(const uint64_t *)(x))
+#endif
 #define MAX_CLOCK_CYCLES     UINT64_MAX
 
+struct dmi_header {
+    uint16_t handle;
+    uint8_t  type;
+    uint8_t  length;
+    uint8_t *data;
+};
+
+/// Global variable holding the cpu frequency.
+uint64_t cpu_f_hz;
+
+static void *pka_cpu_file_read(size_t *len, const char *fname)
+{
+    size_t   rd_cnt = 0;
+    ssize_t  rd_st  = 1;
+    struct   stat sbuf;
+    uint8_t *f;
+    int      fd;
+
+    if ((fd = open(fname, O_RDONLY)) == -1) {
+        if (errno)
+            printf("\n Opening file failed with error:%d.\n", errno);
+        return NULL;
+    }
+
+    if (fstat(fd, &sbuf) == 0) {
+        *len = sbuf.st_size;
+    }
+
+    if ((f = malloc(*len)) == NULL) {
+        printf("\n Memory alloc failed when reading file.\n");
+        goto out;
+    }
+
+    while (rd_cnt != *len && rd_st != 0) {
+        rd_st = read(fd, f+rd_cnt, *len - rd_cnt);
+        if (rd_st == -1) {
+            printf("\n Reading file failed.\n");
+            goto err;
+        }
+        else
+            rd_cnt += rd_st;
+    }
+
+    if (rd_cnt != *len) {
+        printf("\n EOF reached before expected length.\n");
+        goto err;
+    }
+    else
+        goto out;
+
+err:
+    free(f);
+    f = NULL;
+out:
+    if (close(fd) == -1)
+        printf("\n Closing the file failed.\n");
+
+    return f;
+}
+
+static int pka_cpu_checksum(const uint8_t *buf, size_t len)
+{
+    uint8_t sum = 0;
+    size_t  d;
+
+    for (d = 0; d < len; d++)
+         sum += buf[d];
+
+    return sum == 0;
+}
+
+static uint16_t pka_cpu_read_dmi_table(uint32_t len, uint16_t num,
+                               const char *memdev)
+{
+    uint8_t *buf, *data, *next;
+    struct   dmi_header h;
+    size_t   size;
+    int      i = 0;
+
+    if ((buf = pka_cpu_file_read(&size, memdev)) == NULL)
+        return 0;
+
+    len  = size;
+    data = buf;
+
+    while ((i < num || !num)
+        && data + 4 <= buf + len) {

```

Comment:
data+4? it'd be very helpful to define macros.
This may apply to the following as well.